Extract e2e utility code into framework

copejon opened the pull request:
**What this PR does / why we need it**:
There's a growing dependency on Volume e2e utilities related to creating / test against NFS volumes.  For this reason, it's useful to relocate the relevant functions to the `framework` pkg.   Doing so makes these utility functions available to e2e tests outside the `e2e` package.


This PR only moves code from the `e2e` package to `framework` and handle the relevant changes in calls.  It does not change any logic.  


```release-note
NONE
```

@jingxu97 I think there's value here in reducing duplicate code in the `common` package, given that these functions have been copied down to it.  However, there's been some divergence.  Can you PTAL and let me know if there's any reason we can't remove the duplicate `common` code?

cc @jeffvance

spiffxp commented:
Personally I'd like to see `Volume`-related code, test or otherwise, end up owned by @kubernetes/sig-storage-feature-requests.  So eg: move volume related e2e test utils into something like `test/e2e/framework/volume` with a new OWNERS file

Anyone else from @kubernetes/sig-testing-pr-reviews have an opinion on this?

msau42 commented:
I think that sounds reasonable.  There are also many other *_util.go files under framework/, like node, federation, metrics, service, statefulset, etc.  Should each of those also have its own directory?

jeffvance commented on test/e2e/framework/volume_util.go:
```diff
@@ -0,0 +1,379 @@
+	GlusterfsServerImage string = "gcr.io/google_containers/volume-gluster:0.2"
+	CephServerImage      string = "gcr.io/google_containers/volume-ceph:0.1"
+	RbdServerImage       string = "gcr.io/google_containers/volume-rbd:0.1"
+)
```
what do you think about defining a busybox image const here? BB is used a fair amount and always hard-coded.

copejon replied:
good idea, will do

jeffvance commented on test/e2e/framework/volume_util.go:
```diff
@@ -0,0 +1,379 @@
+	Expect(err).NotTo(HaveOccurred())
+}
+
+func DeleteCinderVolume(name string) error {
```
should this (at some point) be moved to a cinder-specific file?

jeffvance commented:
I'm not an assignee but I would approve this. I think it contributes to better volume modularization for the e2e tests. Also re @spiffxp 's comment about adding a _e2e/volume_ dir, this pr gets us closer to being able to do that by eliminating the circular imports we get today if we attempt to add a dir under e2e which relies on _volumes.go_.

jeffvance commented:
@copejon I'd like to see existing code that has replicated some of the func in  _volumes.go_ added to this pr. Eg. _e2e/common/volumes.go_

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42363)
<!-- Reviewable:end -->

spiffxp commented:
@jeffvance I'm not intimately familiar with the framework codebase, so I'm going to defer to someone else for an LGTM, but this otherwise seems uncontroversial to me.

>There are also many other *_util.go files under framework/, like node, federation, metrics, service, statefulset, etc. Should each of those also have its own directory?

@msau42 This sounds reasonable to me, but my intent is not to blow scope on this PR.  I'd prefer to see just this change for volumes to start the trend, or a followup issue filed to account for this.  Ping here with the approach you choose to take and I'll drop an approved accordingly.

jeffvance commented:
@k8s-bot test this

copejon commented:
@spiffxp That's a reasonable plan.  To start, I'll push down the volume util code into a`../framework/volume/` pkg with an `OWNERS` file.   I also agree that it's too much for 1 PR to attempt this for all the util code.

jeffvance commented:
@copejon What do you think about deleting the funcs (eg. in _common/volumes.go_) which are duplicates of funcs in the original _volumes.go_?

jeffvance commented on test/e2e/volumes.go:
```diff
@@ -76,7 +75,6 @@ func DeleteCinderVolume(name string) error {
-
 // These tests need privileged containers, which are disabled by default.  Run
 // the test with "go run hack/e2e.go ... --ginkgo.focus=[Feature:Volumes]"
 var _ = framework.KubeDescribe("Volumes [Feature:Volumes]", func() {
```
Can you change the describe text to "Volumes [Volume]"? Reason:
```
grep -Rn "Feature:Volumes" test/
test/e2e/common/volumes.go:345:// the test with "go run hack/e2e.go ... --ginkgo.focus=[Feature:Volumes]"
test/e2e/volumes.go:385:// the test with "go run hack/e2e.go ... --ginkgo.focus=[Feature:Volumes]"
test/e2e/volumes.go:386:var _ = framework.KubeDescribe("Volumes [Feature:Volume]", func() {
```
grep -Rn "\[Volume\]" test/ |wc -l
120
```

copejon replied:
I can move the `[Volume]` label up from the tests within the suite to the describe line.  `[Feature: ... ]` is used by the CI to identify tests with non-default requirements and could cause side effects if removed:

>  // These tests need privileged containers, which are disabled by default.  Run
 // the test with "go run hack/e2e.go ... --ginkgo.focus=[Feature:Volumes]"

jeffvance replied:
Ok, I remember reading that. Still there is only one occurrence of "Feature:Volumes" in all _test/*_, so it's either not being used (much), or there are no e2e's testing volumes that have non-default requirements. For your tests what are the non-default requirements?

copejon replied:
The `[Feature:]` tag was present when I introduced my tests.  Looking through spec comments, it's required for at least the iSCI test so removing it will likely cause problems.

jeffvance commented:
/assign @jeffvance

jeffvance commented:
/approve

spiffxp commented:
@jeffvance I'm happy to approve/lgtm if you're ok with this, looks like you still had a nit you want fixed

jeffvance commented:
@spiffxp Thanks! I made additional comments today and will wait for Jon to fix or reject. I also assigned myself to this pr so I think that means I can do lgtm when ready.

spiffxp commented:
sgtm will wait the same, suspect my labels would be removed if a new commit was added

jeffvance commented:
/lgtm

jeffvance commented:
/approve
/lgtm

copejon commented:
@k8s-bot ok to test

jeffvance commented:
@k8s-bot ok to test

jeffvance commented:
/approve
/lgtm

jeffvance commented:
@k8s-bot ok to test

copejon committed:
Extract volume test utils into framework.

Moved remaining util functions

moved cinder specific function back to volumes.go, will have to be extracted later when a cinder e2e package is created.

remove dupe code from common/volume.go

Moved [Volume] tags to KubeDescribe

copejon commented:
Something broke in `common/volumes.go`. Debugging it.

jeffvance commented:
/approve
/lgtm

jeffvance commented:
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot non-cri node e2e test this
@k8s-bot node e2e test this

copejon committed:
Add skipUnlessNodeOS is GCI

copejon committed:
Re-add skip unless OS is gci to GCP Volume suite

jeffvance commented:
/approve
/lgtm

copejon commented:
@spiffxp ptal for approve

copejon commented:
ping @timstclair can you ptal && approve?
This PR accomplishes some of what you're aiming for in extracting utility code into the framework package.  This will enable the moving of storage e2e tests into their own package by fixing the circular import problem that current dependencies create.
